### PR TITLE
Add initial TorchScript module importer

### PR DESCRIPTION
**Pull request by silvasean**
It turns out that this was easiest to structure as a general IValue
importer, since torch module are just one of the possible IValue's.

We import the IValue object graph in a braindead fashion into basicpy
ops and a new `torch.nn_module` op that is used to model the
attributes/methods of a torch::jit::Module IValue. See `Torch/ops.mlir`
for an example, and also check out the .py import tests in
`frontends/pytorch/test/module_import`.

As part of this change, a few housekeeping tasks:
- extract some helpers from graph_importer.cpp
- more helpers around the C API
- misc touchups

### CODE REVIEW

**Comment by silvasean** [comment]
@stellaraccident thoughts on the cmake issue?

**Comment by stellaraccident** [comment]
> @stellaraccident thoughts on the cmake issue?

Usually its a bad include path thing (seems unlikely in this case) or a missing dependency on generated sources. Tends to show up in the CI because it hardly has any parallelism. If that's the case, usually running ninja with 1 job max shows the problem. If you are only failing on the CI and can't figure it out, just merge and pitch it back to me to fix.

File: frontends/pytorch/csrc/builder/ivalue_importer.cpp [hunk 1]
```diff
@@ -0,0 +1,177 @@
...
+  void importMethod(torch::jit::Function *function, MlirBlock nnModuleBody);
+
+  MlirBlock importBlock;
+  MlirContext context;
+  TypeMapper typeMapper;
+};
+} // namespace
+
+MlirValue IValueImporter::importModule(torch::jit::Module currentModule) {
+  // TODO: Can we do better?
```
Comment (stellaraccident) [inline frontends/pytorch/csrc/builder/ivalue_importer.cpp:56]: See GraphImporter::extractCallstackLoc(). Probably needs to be lifted to your new common utils. Fine to do in a follow-up.

Also, I've found that I've had to fiddle with this in the past. Not all nodes in TorchScript have sensical locations and it required some probing.

File: frontends/pytorch/test/module_import/primitives.py [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+    def __init__(self):
+        super().__init__()
+        self.i = 3
+        self.f = 42.5
+
+# CHECK: %[[TRUE:.*]] = basicpy.bool_constant true
+# CHECK: %[[N3:.*]] = basicpy.numeric_constant 3 : i64
+# CHECK: %[[N42:.*]] = basicpy.numeric_constant 4.250000e+01 : f64
+# CHECK: %[[MODULE:.*]] = torch.nn_module  {
+# Note: for some reason, Torch always adds a "training" property to all modules.
```
Comment (stellaraccident) [inline frontends/pytorch/test/module_import/primitives.py:24]: Oh, ML frameworks. So predictable.

**Review by stellaraccident — APPROVED** [review]
This is really nice. Great to see things composing.

File: frontends/pytorch/csrc/builder/ivalue_importer.cpp (hunk 1 above)
Comment (silvasean) [reply frontends/pytorch/csrc/builder/ivalue_importer.cpp:56]: The problem is that torch::jit::Module's are not Node's. I guess I could pick a random Node from one of the methods of the module (if there even is a method) and at least get a file name out of it. That seems unsatisfactory though. I would prefer to leave it as-is and wait until we see what kinds of diagnostics we need to emit on these to see what is useful.

**Comment by silvasean** [comment]
> > @stellaraccident thoughts on the cmake issue?
> 
> Usually its a bad include path thing (seems unlikely in this case) or a missing dependency on generated sources. Tends to show up in the CI because it hardly has any parallelism. If that's the case, usually running ninja with 1 job max shows the problem. If you are only failing on the CI and can't figure it out, just merge and pitch it back to me to fix.

Oh, hadn't even seen that issue. I was looking at the failure to `import torch`. Is it intended that this buildbot isn't testing Torch?

```
Traceback (most recent call last):
  File "<string>", line 1, in <module>
ModuleNotFoundError: No module named 'torch'
-- Unable to 'import torch' with /usr/bin/python3 (fallback to explicit config)
CMake Warning at cmake/modules/ConfigurePyTorch.cmake:8 (find_package):
  By not providing "FindTorch.cmake" in CMAKE_MODULE_PATH this project has
  asked CMake to find a package configuration file provided by "Torch", but
  CMake did not find one.

  Could not find a package configuration file provided by "Torch" (requested
  version 1.8) with any of the following names:

    TorchConfig.cmake
    torch-config.cmake

  Add the installation prefix of "Torch" to CMAKE_PREFIX_PATH or set
  "Torch_DIR" to a directory containing one of the above files.  If "Torch"
  provides a separate development package or SDK, be sure it has been
  installed.
Call Stack (most recent call first):
  CMakeLists.txt:174 (NpcompFindPyTorch)
```

**Comment by stellaraccident** [comment]
Just complete lack of time to get it working :(

**Pull request merged**